Examples/test-suite/apply_strings.i

```diff
@@ -31,7 +31,13 @@
 
 %typemap(freearg) SWIGTYPE * ""
 %apply SWIGTYPE* {CharPtr};
+#if defined(SWIGFORTRAN)
+/* Don't try to make a const char* look like char* because of
+ * const correctness. */
+%apply const SWIGTYPE* {CCharPtr};
```

Comment:
I'm confused: that type is defined as
```c
typedef const char* CCharPtr;
```
The original SWIG command (ignored by the `#if` in question) causes an error in the Fortran wrapper code:
```swig
%apply SWIGTYPE* {CCharPtr};
```
which says to treat that constant character pointer as a *mutable* pointer. This seems like an error in the SWIG input, one which the Fortran library catches because it uses some level of const correctness (via `%static_cast` and the like) rather than C-casting everything like most of the other languages.

Is it really the intent in this example code to allow implicitly constant C strings to be treated as a mutable pointer? Fixing this test would require adding an extra `%const_cast` to the `SWIGTYPE*` typemap, which I thought was silly because in no other case but this contrived test is it needed.

Reply:
You probably need to spend a long weekend reading the SWIG manual! Please see http://swig.org/Doc3.0/SWIG.html#SWIG_nn13 regarding how const is handled. I see you have a whole lot of `const SWIGTYPE *` typemaps, they should be removed/changed to `SWIGTYPE *` and then the testcase is likely to work. If you intend to do something different w.r.t. this handling of const (that is uniformly applied across all languages), you'll need to justify this. Does Fortran have a direct equivalent to C's const that can be mirrored in Fortran?

Note that `char *` (both const and non-const) is by default handled as a immutable string in the target language. This test reverts that default to handle it as non-const pointer to a non-const char, which is the default handling for every other pointer type.

Reply:
Haha, it would indeed be a long weekend! Yes, I have read the manual (excluding some of the target languages that are irrelevant to Python and static languages) and vividly remember the const correctness section:
> Although SWIG knows how to correctly deal with const in its internal type system and it knows how to generate wrappers that are free of const-related warnings, SWIG does not make any attempt to preserve const-correctness in the target language.
What I meant when I said "adding an extra `%const_cast` to the `SWIGTYPE*` typemap" is about const-correctness in the *wrapper* code and not in the target language. Most SWIG typemaps avoid const-related errors by using C-style casts rather than `%static_cast`; I was trying to be careful and C++ish in the Fortran typemaps. If you think this is unnecessary I'll simplify the typemaps.

But you bring up an excellent point that I realize is a deviation from the rest of SWIG. Unlike most target languages, Fortran _does_ have qualifiers that are roughly equivalent to C/C++ `const`. Declaring a variable `intent(in)` is roughly the same as a const reference in C, and declaring a variable `intent(out)` means the function argument *must* be an actual variable as opposed to a literal value.

With the wrapping that is in place, the function
```c++
void add_three(int& value);
```
will be declared as `intent(inout)`, which will cause a compiler error if the user tries to write
```fortran
call add_three(4)
```

If the qualifier is omitted, then the above code will compile but generate a run-time fault:
```
Program received signal SIGBUS: Access to an undefined portion of a memory object.
```

I am of the opinion that the additional error catching and self-documenting nature of the generated fortran signatures is worth the extra complexity in the fortran library typemap; but what I could do is remove (most of) the `const` wrapper code specializations and leave the special `imtype` and `ftype` Fortran specializations.

Regardless of that potential change, I now realize that the special `intent` treatment and attempts to be const-correct are not clearly documented in the manual, so if you decide it's OK to leave this feature in, I'll be sure to add a section explicitly describing the behavior and differentiating it from the other languages.

Reply:
This by the way has been addressed in the latest Fortran fork: the unit test no longer requires any special exception for Fortran.